String as integer

jordanmontt opened the pull request:
This PR Does the following:

- Changed the semantics of the method `String>>#asInteger` Now, it returns a integer ONLY if the complete String is a valid integer, else it raises an exception.
- Deprecated `String>>#asSignedInteger`.
- Copied the old implementation of `String>>#asInteger` to  `String>>#extractIntegerPart`. That new method behaves exactly as the old `asInteger`.
- Changed the semantics of `String>>#asUnsignedInteger` to just call `String>>#asInteger` and then getting the absolute value.
- Updated all senders of `String>>#asInteger` to call `String>>#extractIntegerPart` instead.
- Updated and added test to document this behevior.

This also closes #17277

jordanmontt committed:
Added extractIntegerPart method. Changed asInteger, asUnsignedInteger and asSignedInteger implementations to return error when the int cannot be parsed.

jordanmontt committed:
Improved tests

jordanmontt committed:
Replaced senders asInteger -> extractIntegerPart

jordanmontt committed:
Replaced senders asInteger -> extractIntegerPart

jordanmontt committed:
Replaced senders asInteger -> extractIntegerPart

jordanmontt committed:
Moved deprecated method to the deprecated package

jordanmontt committed:
Improved error message when failin to parse.

jordanmontt committed:
Improved method comment

MarcusDenker commented:
Full image load fails with:

```
MetacelloNotification: Loaded -> Calypso-SystemPlugins-Monticello-Browser --- tonel:///builds/workspace/est_and_branch_Pipeline_PR-17299/64/src()
MetacelloNotification: Loaded -> Deprecated13 --- tonel:///builds/workspace/est_and_branch_Pipeline_PR-17299/64/src()
Removing credential.Error: The string '1. 338 forks on GitHub' is not a valid integer
```

which looks related

jordanmontt commented:
Yes, the error is related. It is caused by Microdown. There is already a PR that will fix it: https://github.com/pillar-markup/Microdown/pull/901

But, the Microdown PR requires this PR to be merged first. So, We will have the failing tests for a build, then we merge the Microdown PR and all the tests should return to be green.

astares commented on src/Collections-Strings-Tests/StringTest.class.st:
```diff
@@ -447,11 +447,11 @@ StringTest >> testAsHTMLString [
-	self assert: 'abc234def567' asInteger equals: 234.
+	self should: [ '1796exportFixes-tkMX' asInteger ] raise: Error.
+	self should: [ 'donald' asInteger ] raise: Error.
+	self should: [ 'abc234def567' asInteger ] raise: Error.
```
Please add a case 

```Smalltalk
'123 4' asInteger 
```
with a space as this shows clearly what can happens if for instance in an edit field someone uses accidentially a blank. In existing Pharo 12 this is (unfortunately) accepted and returns 123 without an exception.

jordanmontt committed:
Added '123 4' asInteger  test case as suggested by astares

jordanmontt commented:
Thanks for the suggestion @astares   I added the test case

Ducasse commented:
I have no idea how to fix these tests. Probably somewhere in the baseline the microdown inside Pharo should be unloaded
but I have no idea where it is. 

I will propose to remove Microdown from Pharo because it just slows me down all the time. 
For the record all microdown tests are green and I'm systematically adding more of them. 

```
#########################
# 2 tests did not pass: #
#########################

ReleaseTest
 ✗ #testUndeclared (151ms)
CollectionIsEmpty: a WeakIdentitySet() is empty
WeakIdentitySet(Collection)>>errorEmptyCollection
WeakIdentitySet(Collection)>>emptyCheck
WeakIdentitySet(Collection)>>anyOne
WeakIdentitySet(WeakSet)>>anyOne
[ :variable |
			               | method |
			               method := variable usingMethods anyOne.
			               stream cr
				               nextPutAll: variable name;
				               nextPutAll: ' in: ';
				               print: method methodClass;
				               nextPutAll: '>>';
				               print: method selector ] in [ :stream |
		               stream nextPutAll: 'Found undeclared Variables: '.
		               remaining do: [ :variable |
			               | method |
			               method := variable usingMethods anyOne.
			               stream cr
				               nextPutAll: variable name;
				               nextPutAll: ' in: ';
				               print: method methodClass;
				               nextPutAll: '>>';
				               print: method selector ] ] in [ "we compile a second method with the undeclared #undeclaredStubInstVar1 to trigger the code path of removing twice in #cleanOutUndeclared"
	self class compile: 'methodForTest ^undeclaredStubInstVar1'.
	Smalltalk cleanOutUndeclared.
	undeclaredVariables := Undeclared associations select: [ :each | each isUndeclaredVariable ].

	validExceptions := { #undeclaredStubInstVar1. #undeclaredStubInstVar2. #Gofer }. "The laste one is for Smalltalk CI and our external projects... But we should find a better solution at some point."

	"for now we filter by name, maybe filtering by variable would be better"
	remaining := undeclaredVariables reject: [ :each | validExceptions includes: each name ].

	"we look for one of the using methods of the undeclared var and report that,
	this should be enough to fix it quickly"
	description := String streamContents: [ :stream |
		               stream nextPutAll: 'Found undeclared Variables: '.
		               remaining do: [ :variable |
			               | method |
			               method := variable usingMethods anyOne.
			               stream cr
				               nextPutAll: variable name;
				               nextPutAll: ' in: ';
				               print: method methodClass;
				               nextPutAll: '>>';
				               print: method selector ] ].

	self assert: remaining isEmpty description: description ] in ReleaseTest>>testUndeclared ...anyOne
Array(SequenceableCollection)>>do:
[ :stream |
		               stream nextPutAll: 'Found undeclared Variables: '.
		               remaining do: [ :variable |
			               | method |
			               method := variable usingMethods anyOne.
			               stream cr
				               nextPutAll: variable name;
	description := String streamContents: [ :stream |
		               stream nextPutAll: 'Found undeclared Variables: '.
		               remaining do: [ :variable |
			               | method |
			               method := variable usingMethods anyOne.
			               stream cr
				               nextPutAll: variable name;
				               nextPutAll: ' in: ';
				               print: method methodClass;
				               nextPutAll: '>>';
				               print: method selector ] ].

	self assert: remaining isEmpty description: description ] in ReleaseTest>>testUndeclared ...streamContents: [ :stream |
		               stream nextPutAll: 'Found undeclared Variables: '.
		               remaining do: [ :variable |
			               | method |
			               method := variable usingMethods anyOne.
			               stream cr
				               nextPutAll: variable name;
				               nextPutAll: ' in: ';
				               print: method methodClass;
				               nextPutAll: '>>';
				               print: method selector ] ]
FullBlockClosure(BlockClosure)>>ensure:
ReleaseTest>>testUndeclared ...ensure: [ self class removeSelector: #methodForTest ]
ReleaseTest(TestCase)>>performTest
 ✗ #testObsoleteClasses (831ms)
TestFailure: Obsolete classes remaining: an Array(AnObsoleteMicResourceReference AnObsoleteMicAbsoluteResourceReference AnObsoleteMicHTTPResourceReference AnObsoleteMicPharoImageResourceReference AnObsoleteMicrodownVisitor AnObsoleteMicArgumentList AnObsoleteMicMicrodownSharedPool AnObsoleteMicAbstractDelimiter AnObsoleteMicAnchorReferenceCloserDelimiter AnObsoleteMicAnchorReferenceOpenerDelimiter AnObsoleteMicAnnotationCloserDelimiter AnObsoleteMicAnnotationOpenerDelimiter AnObsoleteMicBoldDelimiter AnObsoleteMicFigureNameOpenerDelimiter AnObsoleteMicItalicDelimiter AnObsoleteMicLinkNameDelimiter AnObsoleteMicLinkNameCloserDelimiter AnObsoleteMicLinkNameOpenerDelimiter AnObsoleteMicMathDelimiter AnObsoleteMicMonospaceDelimiter AnObsoleteMicRawCloserDelimiter AnObsoleteMicRawOpenerDelimiter AnObsoleteMicStrikeDelimiter AnObsoleteMicElement AnObsoleteMicAbstractBlock AnObsoleteMicAbstractAnnotatedBlock AnObsoleteMicAnnotatedBlock AnObsoleteMicContinuousMarkedBlock AnObsoleteMicCommentBlock AnObsoleteMicQuoteBlock AnObsoleteMicTableBlock AnObsoleteMicListBlock AnObsoleteMicOrderedListBlock AnObsoleteMicUnorderedListBlock AnObsoleteMicListItemBlock AnObsoleteMicParagraphBlock AnObsoleteMicRootBlock AnObsoleteMicSingleLineBlock AnObsoleteMicAnchorBlock AnObsoleteMicHeaderBlock AnObsoleteMicHorizontalLineBlock AnObsoleteMicStartStopMarkupBlock AnObsoleteMicEnvironmentBlock AnObsoleteMicMetaDataBlock AnObsoleteMicSameStartStopMarkupBlock AnObsoleteMicAbstractCodeBlock AnObsoleteMicScriptBlock AnObsoleteMicPharoEvaluatorBlock AnObsoleteMicMathBlock AnObsoleteMicInlineElement AnObsoleteMicAnnotationBlock AnObsoleteMicInlineBlockWithUrl AnObsoleteMicFigureBlock AnObsoleteMicLinkBlock AnObsoleteMicTextBlock AnObsoleteMicInlineParser AnObsoleteMicrodownParser AnObsoleteMicRichTextBrush AnObsoleteMicRichTextIndentBrush AnObsoleteMicRichTextCanvas AnObsoleteMicRichTextCodeBlockStyler AnObsoleteMicTextStyler AnObsoleteMicRichTextComposer AnObsoleteMicRichTextDoIt AnObsoleteMicMorphicTextAdapter AnObsoleteMicMorphicMicrodownAdapter AnObsoleteMicScrolledTextMorph)
ReleaseTest(TestAsserter)>>assert:description:resumable:
ReleaseTest(TestAsserter)>>assert:description:
ReleaseTest>>testObsoleteClasses ...assert: obsoleteClasses isEmpty
		description: [
			String streamContents: [ :s|
				s
					nextPutAll: 'Obsolete classes remaining: ';
					print: obsoleteClasses ]]
ReleaseTest(TestCase)>>performTest


###########
# Summary #
###########

ReleaseTest
 ✗ #testUndeclared (151ms)
 ✗ #testObsoleteClasses (831ms)


  Executed 765 Tests with 1 Failures and 1 Errors in 62.56s.
```

jordanmontt commented:
Hello @Ducasse 

All the tests are now green for Microdown, yes. The problem is that there are 2 tests that *will* fail when we will integrate this PR, because they depend on the old `asString`.

What I propose is:
1) integrate this PR. This will cause the Pharo CI to have 2 failing tests because of Microdown.
2) Integrate the PR in Microdown https://github.com/pillar-markup/Microdown/pull/901 This will fix the broken tests for Microdown
3) Update Pharo to use latests microdown
4) Everything, both Pharo and Microdown will be green at this point.

guillep committed:
Fix closing window shortcuts in combination with window groups

MarcusDenker committed:
Merge branch 'Pharo13' into fix-window-closing

Gabriel-Darbord committed:
Fix: Rubric error when jumping to previous word and string begins with separators

tinchodias committed:
OSWindow-SDL2: setCursor if window is focused and when focus is gained

Fixes #18234

tinchodias committed:
Revert wrong deletion of 2 methods

MarcusDenker committed:
fixes #18225, adds empty method as described in the issue

jecisc committed:
Merge pull request #18238 from MarcusDenker/18225-13-Be-robust-against-unintentionally-sending-initialize-to-a-Trait

Fix for #18225: Be robust against unintentionally sending #initialize to a Trait (Pharo 13)

jecisc committed:
Merge pull request #18232 from Gabriel-Darbord/P13-fix-rubric-previous-word-separators-at-start

[P13 Backport] Fix: Rubric error when jumping to previous word and string begins with separators

tesonep committed:
Adding validation of the hierarchy for all the slots
Fix #18227

tesonep committed:
Supporting the case when the hierarchy is deeper.

tesonep committed:
Adding exceptions to QA rules

tesonep committed:
Using the instance variable so the check of unused instance variables does not complain

tesonep committed:
Merge branch 'Pharo13' into fix-window-closing

tesonep committed:
Merge pull request #18243 from tesonep/18227-How-to-break-a-hierarchy-P13

[Pharo13] Adding validation of the hierarchy for all the slots

tesonep committed:
Merge branch 'Pharo13' into fix-window-closing

tesonep committed:
Merge pull request #18116 from guillep/fix-window-closing

Fix closing window shortcuts in combination with window groups

balsa-sarenac committed:
Fix: cleanup undeclared registry after renaming a class

tesonep committed:
Updating Roassal to a latest release fixing some issues of the menu entries

tesonep committed:
Fixing the Toplo Version

tesonep committed:
Merge pull request #18263 from pharo-project/fixing-toplo-version

Fixing the Toplo Version

tesonep committed:
Merge branch 'Pharo13' into change-roassal-version

MarcusDenker committed:
Merge pull request #18262 from tesonep/change-roassal-version

[Pharo13] Update Roassal to new Released version

MarcusDenker committed:
Merge pull request #18280 from tinchodias/FixCursorBackport

Fix cursor on multiple OSWindows

jecisc committed:
Fix Duplicate class refactoring if there are spaces in package or package tag

Currently StRequestClassPresenter is checking if the name of a class, package and package tag are right. But is expects that we have no space in the package name or package tag name. But it is possible to have spaces there in Pharo!

Update the check.

Fixes #18287

jecisc committed:
Fix protocol

jecisc committed:
Add tests on class package and tag name validation

jecisc committed:
Merge pull request #18290 from jecisc/refactorings/backport

Backport #18289 to P13: Fix Duplicate class refactoring

demarey committed:
rename examples and ClapHelp command

Ducasse committed:
Merge pull request #18312 from demarey/clap-help-fix

rename examples and ClapHelp command

estebanlm committed:
updated spec to v2.0.3 and newtools to v0.10.4

estebanlm committed:
Merge pull request #18313 from estebanlm/Pharo13

uodate new tools to v0.10.4 and spec to v2.0.3

jecisc committed:
Fix cancelling of FilePresenter via UIManager

If you request a filename via the UIManager and cancel, the name is still returned. This causes bugs. For example, when cancelling a "Save as...", it is still saved.
I also removed a hardcoded ref to MorphicUIManager.

Fixes #18221

balsa-sarenac committed:
Fix: rename method violators are list and not single instance

balsa-sarenac committed:
Fix: use correct method to find method

Ducasse committed:
Merge pull request #18363 from balsa-sarenac/fix/rename-method-violators

Fix: rename method violators are list and not single instance

guillep committed:
Fix lock DNU on P12 and on

tesonep committed:
Patching the process so the files are P30 and not P31

tesonep committed:
- Adding restart message to the UIManager to force the restart, restarting Morphic or the UI handler that we are using
- Adding a menu option in MacOS for restarting the UIManager

Rinzwind committed:
Updated version in #iceberg for BaselineOfPharo to ‘v2.4.2’.

tesonep committed:
- Adding event to handle beforeMainPharoWindowClosed
- Adding operation to the OSWorldRenderer to restart
- Adding a menu entry for the main window (in Windows) to reset the OSWorldRenderer
- Better handling of SystemWindowEvents
- Get information about the RemoteTerminal
- Handle logon of RemoteDesktop to force the OSWorldRenderer restarting

tesonep committed:
Hardcodding version for the tests

MarcusDenker committed:
Merge pull request #18379 from tesonep/continue-to-generate-P30-artifacts

continue-to-generate-P30-artifacts

MarcusDenker committed:
Merge branch 'Pharo13' into fix/backport-rename-and-undeclared-registry

MarcusDenker committed:
Merge pull request #18368 from balsa-sarenac/fix/backport-rename-and-undeclared-registry

Backport: fix cleanup undeclared registry after renaming a class

as this is a backport and the unix ci is not related, I merge

tesonep committed:
Merge branch 'Pharo13' into fix-metacello-lock-13

tesonep committed:
Merge pull request #18394 from guillep/fix-metacello-lock-13

Fix lock DNU on P13 (port #18377)

tesonep committed:
Merge branch 'Pharo13' into P13-adding-restart-world-option

tesonep committed:
Merge pull request #18396 from tesonep/P13-adding-restart-world-option

P13-adding-restart-world-option

guillep committed:
Remove tests of deprecated code raising useless warnings

jecisc committed:
Merge branch 'Pharo13' into fix-deprecated-tests

Ducasse committed:
Merge pull request #18414 from guillep/fix-deprecated-tests

Remove tests of deprecated code raising useless warnings

Rinzwind committed:
Merge branch 'Pharo13' into update-baseline-pharo13-iceberg

jecisc committed:
Merge pull request #18388 from Rinzwind/update-baseline-pharo13-iceberg

Update version in #iceberg for BaselineOfPharo for Pharo 13 to ‘v2.4.2’

tesonep committed:
Fix #18431 for P13

We should destroy ZdcPluginSSLSessions when they are not used anymore.
Just in case the ZnClient has been not closed correctly. Adding a test

tesonep committed:
Merge branch 'Pharo13' into 18431-P13-ZdcPluginSSLSession-leaks-SSLSession-information-if-the-connection-is-not-explicitly-closed

tesonep committed:
Removing duplicated code in the scripts

tesonep committed:
Fixing upload script

tesonep committed:
Merge branch 'Pharo13' into 18431-P13-ZdcPluginSSLSession-leaks-SSLSession-information-if-the-connection-is-not-explicitly-closed

tesonep committed:
Merge pull request #18434 from tesonep/18431-P13-ZdcPluginSSLSession-leaks-SSLSession-information-if-the-connection-is-not-explicitly-closed

[P13] 18431-ZdcPluginSSLSession-leaks-SSLSession-information-if-the-connection-is-not-explicitly-closed

labordep committed:
Fix auto refactoring

jecisc committed:
Merge pull request #18437 from labordep/Pharo13

Fix auto refactoring

Ducasse commented:
Hi Seb
I'm about to make a new version of microdown. Here is what I will do. I will introduce a MicrodownOnlyUser extension of your definition of asInteger and use it in Microdown. This way nothing will break.
Then once this PR is integrated I will just migrate to your version.

jordanmontt committed:
Merge branch 'Pharo13' into string-as-integer

Ducasse commented:
Hi cyril 
it is also related to this issue but I have no idea what to do to address the conflict
https://github.com/pillar-markup/Microdown/issues/958
If you get an idea.
Basically in MD I just introduced a private method so that asInteger can be changed. I think that this is nice to have also extractNumber

Now I do not get the problem with the gitbridge move.

jecisc committed:
Fix behavior of FileBrowser when we cancel the selection

The UITheme was not managing the cancel button of the FileBrowser. Here is a fix even if ideally we should not use the UITheme anymore in the future for this.

Fixes https://github.com/pharo-spec/NewTools/issues/1230

jecisc committed:
Merge pull request #18478 from jecisc/backport/uitheme-cancel

[Backport] Manage cancelling of FileBrowser via UITheme

jordanmontt committed:
Merge branch 'Pharo13' into string-as-integer

jecisc committed:
Allow to set tags priority in Iceberg

MarcusDenker committed:
Merge pull request #18499 from pharo-project/jecisc-patch-1

Allow to set tabs priority in Iceberg

jordanmontt commented:
We need, in Microdown, to merge dev into Phar14. There has been fixed in Pharo14 branch that normally should fix this PR

jordanmontt committed:
Merge branch 'Pharo13' into string-as-integer

jordanmontt committed:
Added comment

jordanmontt committed:
Merge remote-tracking branch 'upstream/Pharo14' into string-as-integer

jordanmontt committed:
Merge branch 'Pharo14' into string-as-integer

jordanmontt committed:
Revert "Merge branch 'Pharo14' into string-as-integer"

This reverts commit b16d87fd3b7084915b5a853c06506413a0fe7e81, reversing
changes made to 109fdf4c00eb8999fda7c04cd5db443f99b8d279.
